<h1><center> TP2 : Pandas, Librairie d'analyse de données </center></h1>

Le but de ce TP est de se familiariser avec la classe 'pandas' de Python. Cette bibliothèque est souvent utilisée pour l'analyse des données et est pratique pour les outils de manipulation.
Nous considérons une étude de cas basée sur un ensemble de données contenant des informations sur les réservations de deux hôtels. L'Hôtel 1 est un complexe hôtelier et l'Hôtel 2 est un hôtel de centre-ville au Portugal. L'ensemble de données a été publié par https://www.sciencedirect.com/science/article/pii/S2352340918315191.

Nous listons d'abord les fonctions de base de pandas. La PARTIE 1 vise à utiliser pandas comme outils de visualisation pour une meilleure compréhension des données. La PARTIE 2 montre à quel point il est facile de combiner des dataframes "pandas" et des modèles "sklearn" pour créer des fonctionnalités supplémentaires et prévoir. 


En guise de devoirs, nous vous proposons un problème très concret qui est ouvert et pour lequel nous attendons votre créativité (comme d'habitude)! 

In [2]:
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib notebook

# Petite intro à pandas

In [3]:
data = pd.read_csv("bookings.csv") 

In [4]:
data.head(5) # imprimer les 5 premières entrées des données

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

In [5]:
data.columns # donne les noms des colonnes

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

Cet ensemble de données contient des informations sur les réservations de deux hôtels. L'Hôtel 1 est un complexe hôtelier et l'Hôtel 2 est un hôtel de centre-ville au Portugal. L'ensemble de données a été publié par https://www.sciencedirect.com/science/article/pii/S2352340918315191.

Ces données ont 32 columns :

1. **hotel** -- un des deux hotels
2. **is_canceled** -- valeur indiquant si la réservation a été annulée (1) ou non (0) 
3. **lead_time** -- nombre de jours écoulés entre la date d'entrée de la réservation dans le PMS et la date d'arrivée 
4. ....

Pour une description complète de chaque colonne, veuillez consulter : https://www.kaggle.com/jessemostipak/hotel-booking-demand

In [6]:
data['country'] # on peut aussi voir chaque colonne du jeu de données

0         PRT
1         PRT
2         GBR
3         GBR
4         GBR
         ... 
119385    BEL
119386    FRA
119387    DEU
119388    GBR
119389    DEU
Name: country, Length: 119390, dtype: object

In [8]:
data.shape

(119390, 32)

In [9]:
data['country'].unique() # liste toutes les valeurs présentes de manière unique dans la colonne

array(['PRT', 'GBR', 'USA', 'ESP', 'IRL', 'FRA', nan, 'ROU', 'NOR', 'OMN',
       'ARG', 'POL', 'DEU', 'BEL', 'CHE', 'CN', 'GRC', 'ITA', 'NLD',
       'DNK', 'RUS', 'SWE', 'AUS', 'EST', 'CZE', 'BRA', 'FIN', 'MOZ',
       'BWA', 'LUX', 'SVN', 'ALB', 'IND', 'CHN', 'MEX', 'MAR', 'UKR',
       'SMR', 'LVA', 'PRI', 'SRB', 'CHL', 'AUT', 'BLR', 'LTU', 'TUR',
       'ZAF', 'AGO', 'ISR', 'CYM', 'ZMB', 'CPV', 'ZWE', 'DZA', 'KOR',
       'CRI', 'HUN', 'ARE', 'TUN', 'JAM', 'HRV', 'HKG', 'IRN', 'GEO',
       'AND', 'GIB', 'URY', 'JEY', 'CAF', 'CYP', 'COL', 'GGY', 'KWT',
       'NGA', 'MDV', 'VEN', 'SVK', 'FJI', 'KAZ', 'PAK', 'IDN', 'LBN',
       'PHL', 'SEN', 'SYC', 'AZE', 'BHR', 'NZL', 'THA', 'DOM', 'MKD',
       'MYS', 'ARM', 'JPN', 'LKA', 'CUB', 'CMR', 'BIH', 'MUS', 'COM',
       'SUR', 'UGA', 'BGR', 'CIV', 'JOR', 'SYR', 'SGP', 'BDI', 'SAU',
       'VNM', 'PLW', 'QAT', 'EGY', 'PER', 'MLT', 'MWI', 'ECU', 'MDG',
       'ISL', 'UZB', 'NPL', 'BHS', 'MAC', 'TGO', 'TWN', 'DJI', 'STP',
       'KNA', 'E

In [10]:
data.count()

hotel                             119390
is_canceled                       119390
lead_time                         119390
arrival_date_year                 119390
arrival_date_month                119390
arrival_date_week_number          119390
arrival_date_day_of_month         119390
stays_in_weekend_nights           119390
stays_in_week_nights              119390
adults                            119390
children                          119386
babies                            119390
meal                              119390
country                           118902
market_segment                    119390
distribution_channel              119390
is_repeated_guest                 119390
previous_cancellations            119390
previous_bookings_not_canceled    119390
reserved_room_type                119390
assigned_room_type                119390
booking_changes                   119390
deposit_type                      119390
agent                             103050
company         

Cet ensemble de données contient 119390 réservations différentes. Certaines réservations ont des valeurs manquantes. 


In [11]:
data.values # Un data frame peut être converti en une numpy array:

array([['Resort Hotel', 0, 342, ..., 0, 'Check-Out', '2015-07-01'],
       ['Resort Hotel', 0, 737, ..., 0, 'Check-Out', '2015-07-01'],
       ['Resort Hotel', 0, 7, ..., 0, 'Check-Out', '2015-07-02'],
       ...,
       ['City Hotel', 0, 34, ..., 4, 'Check-Out', '2017-09-07'],
       ['City Hotel', 0, 109, ..., 0, 'Check-Out', '2017-09-07'],
       ['City Hotel', 0, 205, ..., 2, 'Check-Out', '2017-09-07']],
      dtype=object)

In [12]:
data['is_canceled'].values

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

Cependant, ce tableau ne peut pas être directement alimenté dans un modèle scikit-learn.

1. les valeurs sont hétérogènes (des chaînes de caractères pour les catégories, des entiers et des nombres à virgule)
2. certaines valeurs d'attributs sont manquantes 

# Prédiction d'annulations : Partie I -- visualisation

**Nos buts** Le but de cette partie est de fournir quelques exemples de visualisation combinant ```pandas```, ```matplotlib```, ```seaborn```. 

Nous examinerons une tâche très naturelle et pratique - prédire l'annulation d'une réservation donnée. Bien sûr, le premier instinct du 'praticien ML' moderne est de mettre toutes les données sur un réseau de neurone et d'effectuer un tas de réglages fins. Il y a vraiment beaucoup de problèmes avec une telle approche. Règle générale : explorez vos données avant de créer des pipelines ML! Il est toujours plus intéressant d'étudier les données et de trouver des modèles interprétables. 

Pandas permet de manipuler l'ensemble de données d'une manière très pratique. Ceux qui connaissent SQL l'apprécieront certainement!

En tant que première tâche de visualisation, comprenons quels mois ont le plus grand nombre d'annulations.
Nous commençons par créer un nouvel ensemble de données contenant les informations que nous aimerions tracer. 

In [ ]:

'''
 On crée un jeu de données pour chaque hôtel .groupby("arrival_date_month") regroupera les observations par mois
 et la fonction .count() comptera simplement le nombre de réservations pour chaque mois
'''
n_reserv_H1 = data.loc[(data["hotel"] == "Resort Hotel")].groupby("arrival_date_month")["hotel"].count()
n_cancel_H1 = data.loc[(data["hotel"] == "Resort Hotel")].groupby("arrival_date_month")["is_canceled"].sum()

# On fait la même chose pour le deuxième hôtel
n_reserv_H2 = data.loc[(data["hotel"] == "City Hotel")].groupby("arrival_date_month")["hotel"].count()
n_cancel_H2 = data.loc[(data["hotel"] == "City Hotel")].groupby("arrival_date_month")["is_canceled"].sum()

In [ ]:
n_reserv_H1.head() # encore une fois, nous pouvons jeter un oeil aux 5 meilleures entrées 

In [ ]:
n_reserv_H1['April'] # nombre de réservations pour le Resort Hotel en avril 

In [ ]:
n_cancel_H1['April'] # nombre d'annulations pour le Resort Hotel en avril

In [ ]:
# enfin, on rassemble tout 


data_visualH1 = pd.DataFrame({"hotel": "Resort Hotel",
                                "month": list(n_reserv_H1.index),
                                "n_booking": list(n_reserv_H1.values),
                                "n_cancel": list(n_cancel_H1.values)})
data_visualH2 = pd.DataFrame({"hotel": "City Hotel",
                                "month": list(n_reserv_H2.index),
                                "n_booking": list(n_reserv_H2.values),
                                "n_cancel": list(n_cancel_H2.values)})
data_visual = pd.concat([data_visualH1, data_visualH2], ignore_index=True)

# remarquez à quel point il est facile d'ajouter une nouvelle colonne. Nous écrivons simplement ce qui suit 
data_visual["percent_cancel"] = data_visual["n_cancel"] / data_visual["n_booking"] * 100 # pourcentage d'annulations

In [ ]:
data_visual.head() # nos données finales

In [ ]:
# Fournir des graphes est simple une fois que nous avons un bon jeu de données. 
plt.figure(figsize=(6, 6))
sns.barplot(x = "month", y = "percent_cancel" , hue="hotel",
            hue_order = ["Resort Hotel", "City Hotel"], data=data_visual)
plt.title("Cancelations per month")
plt.xticks(rotation=45)
plt.ylabel("Cancelations [%]")
plt.legend()
plt.show()

Le graphe précédent est sympa, mais l'ordre des mois est très ennuyeux!  

**Question 1.** Proposez une solution qui réorganisera le barplot ci-dessus en utilisant l'ordre chronologique mensuelle standard. Astuce: utilisez la fonction ```pd.Categorical()``` de pandas.

**Question 2.** Fournissez une interprétation du graphique ci-dessus. 

**Question 3.** Quel sont les deux pays d'origine qui font le plus de réservations dans chaque hôtel ?

**Question 4.** Indiquez le nombre d'annulations pour les clients répétés et non répétés pour les deux hôtels.  

In [ ]:
plt.figure(figsize=(6, 6))
sns.countplot(x="is_canceled", hue='is_repeated_guest', data=data)
plt.title("Canceled vs repeated guest", fontsize=16)
plt.plot()

La plupart des clients de ces deux hôtels ne sont pas répétés, tandis que les clients répétés sont moins susceptibles d'annuler. 

In [ ]:
data_req = data[(data['hotel'] == 'City Hotel')].groupby(['total_of_special_requests', 'is_canceled']).size().unstack(level=1)
data_req.plot(kind='bar', stacked=True, figsize=(6,6))
plt.title('Special Request vs Cancellation in H2')
plt.xlabel('Number of Special Request', fontsize=10)
plt.xticks(rotation=300)
plt.ylabel('Count', fontsize=10)

La plupart des réservations dans l'hôtel de ville n'ont pas de demandes spéciales et l'annulation dans ce cas est presque de 50/50. Cependant, lorsque des demandes spéciales sont effectuées, le taux d'annulation est nettement inférieur. 

**Question 5.** Faites le même graphe pour l'hôtel de villégiature. Faites vos conclusions.

# Prédiction d'annulations : Partie II -- ML

**Nos buts** Le message principal ici est - ne réinventez pas la roue. Les quelques lignes de code suivantes mettent en évidence la simplicité avec laquelle nous pouvons combiner des dataframes ```pandas``` et des modèles ```sklearn```. En apprenant quelques outils simples (i.e. ```pipeline```, ```gridsearchcv```) notre code devient lisible, compact et peut être utilisé pour créer des fonctionnalités supplémentaires.

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

In [ ]:
numeric_features = ["lead_time", "arrival_date_week_number", "total_of_special_requests"]
categorical_features = ["hotel", "market_segment","deposit_type","customer_type"]
features = numeric_features + categorical_features
X = data.drop(["is_canceled"], axis=1)[features]
y = data["is_canceled"]

Avant d'utiliser un algorithme ML de sklearn, nous devons gérer les valeurs manquantes. Il n'y a pas de réponse unique sur la façon de traiter les valeurs manquantes dans vos données. Nous utiliserons l'approche la plus simple (peut-être). Tout d'abord, si la variable est numérique et est manquante, nous allons la remplacer par zéro. Deuxièmement, si la variable est catégorielle et est manquante, nous allons définir une nouvelle catégorie et l'appeler ```Not defined```.

In [ ]:
numeric_transformer = SimpleImputer(strategy="constant", fill_value=0) # to deal with missing numeric data
categorical_transformer = Pipeline(steps=[
                                    ("imputer", SimpleImputer(strategy="constant", fill_value="Not defined")),
                                    ("onehot", OneHotEncoder(handle_unknown='ignore'))]) # to deal with missing categorical data 
preproc = ColumnTransformer(transformers=[("num", numeric_transformer, numeric_features),
                                          ("cat", categorical_transformer, categorical_features)])

**Question:** Qu'est-ce que ```OneHotEncoder()``` ? Pourquoi l'utilise-t-on dans notre cas ?

Imaginez maintenant une situation où vous voulez essayer beaucoup de modèles différents et pour chaque modèle vous voulez faire une validation croisée pour sélectionner les meilleurs hyperparamètres. En plus de cela, vous souhaitez prétraiter les données afin de fournir quelque chose de significatif dans chaque méthode.

Le morceau de code suivant montre comment le faire. 

In [ ]:
models = [("logreg", LogisticRegression(max_iter=500))]
grids = {"logreg" : {'logreg__C': np.logspace(-2, 2, 5, base=2)}}
for name, model in models:
    pipe = Pipeline(steps=[('preprocessor', preproc), (name, model)])
    clf = GridSearchCV(pipe, grids[name], cv=3)
    clf.fit(X, y)
    print('Results for {}'.format(name))
    print(clf.cv_results_)

**Question:** Dans l'exemple précédent, nous rencontrons à nouveau le problème de convergence. Bien sûr, nous pouvons définir un nombre plus élevé d'itérations, mais cela prend du temps. Comme vous l'avez vu, une normalisation appropriée peut résoudre le problème. Insérez une étape de normalisation dans le pipeline. Notez que nous ne voulons pas normaliser les données catégorielles, cela n'a tout simplement pas de sens. Veillez à ne normaliser que les données numériques. Est-ce que ça a résolu l'avertissement? 

**Question:** Comme nous pouvons le voir, le code précédent n'utilise que la régression logistique. Modifiez le code ci-dessus en insérant votre méthode ML préférée. 

**Remark:** Notez que dans cette partie, nous n'avons sélectionné qu'un petit sous-ensemble de fonctionnalités. Nous aurions pu également utiliser d'autres fonctionnalités. 

# Les devoirs

**L'histoire !**
Vous êtes un data scientist travaillant pour l'hôtel, un jour le manager vient vous dire :

Dans nos hôtels, nous avons la possibilité d'offrir une place de parking, ce qui coûte de l'argent. Apparemment, tous les clients ne sont pas conscients que nous avons une telle option. Je veux vraiment proposer des places de parking à tous ceux qui en ont besoin, car l'hôtel peut gagner plus d'argent comme ça. Écoutez, je peux envoyer une notification par SMS à nos clients. Bien sûr, les SMS ne sont pas gratuits et, de plus, les gens sont très en colère s'ils reçoivent des notifications stupides sans raison. Pour chaque nouvelle réservation, j'aimerais décider si je dois ou non envoyer la notification au client.  

**Problème** Explorez vos données pour aider le manager et construire un algorithme de prédiction, en utilisant le modèle ci-dessus comme source d'inspiration. 

**Attention!** Certaines colonnes ne sont pas réellement disponibles au moment de la réservation. Par exemple, la colonne cible ```is_canceled``` de la partie précédente ne peut clairement pas être observée au moment où nous devons décider d'envoyer le SMS. 